### PR TITLE
[NNVM][TENSORFLOW] LSTM operator and PTB word prediction frontend

### PR DESCRIPTION
**Pull request by joyalbin**
This PR contain:
- [x] LSTM operator implementation and layer based operator parsing in Tensorflow
- [x] Single LSTMBlockCell layer model support
- [x] PTB LSTMBlockCell based model frontend script
- [x] Frontend testcases

### CODE REVIEW

**Comment by joyalbin** [comment]
@Huyuwei @masahi @srkreddy1238 Please help to review this PR

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 1]
```diff
@@ -469,6 +469,224 @@ def _impl(inputs, attr, params):
...
+
+def _infer_out_shapes(inputs, params):
+    """A hack for getting output shape of a node.
+    """
+    g = _graph.create(inputs)
+    shape_dict = {k: v.shape for k, v in params.items()}
+    _, out_shapes = graph_util.infer_shape(g, **shape_dict)
+    return out_shapes
+
+def _stridedSlice():
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Now we have strided_slice operator from nnvm. We should use it.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 2]
```diff
@@ -469,6 +469,224 @@ def _impl(inputs, attr, params):
...
+        new_input.append(inputs.pop(0))
+        new_input.append(inputs.pop(0))
+        return AttrCvt(
+            op_name="take",
+            extras={'axis':axis},
+            ignores=['Tindices', 'Tparams', 'validate_indices', \
+                     'Taxis', '_class'])(new_input, attr)
+    return _impl
+
+def _infer_out_shapes(inputs, params):
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: We have already _input_shapes attribute. Any challenge using it?

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 3]
```diff
@@ -544,3 +711,6 @@ def test_forward_mobilenet():
     test_forward_inception_v1()
     test_forward_mobilenet()
     test_forward_variable()
+    test_forward_lstm()
```
Comment (srkreddy1238) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Suggest to add a real time end to end test case for LSTM.

File: tutorials/nnvm/from_tensorflow_rnn.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
+"""
+Tutorial for Tensorflow RNN Models
```
Comment (srkreddy1238) [inline tutorials/nnvm/from_tensorflow_rnn.py]: underline up to the text length pls.

File: tutorials/nnvm/from_tensorflow_rnn.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+from google.protobuf import text_format
+from tensorflow.python.framework import graph_util
+dir(tf.contrib)
+
+
+sample_repo = 'http://www.fit.vutbr.cz/~imikolov/rnnlm/'
+sample_data_file = 'simple-examples.tgz'
+sample_url = sample_repo+sample_data_file
+
+ptb_repo = 'https://github.com/joyalbin/dmlc_store/raw/master/trained-models/tf/ptb/pb/'
```
Comment (srkreddy1238) [inline tutorials/nnvm/from_tensorflow_rnn.py]: use dmlc/web-data to store these model.

File: tutorials/nnvm/from_tensorflow_rnn.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+def pretty_print(items, is_char_model, id2word):
+    if not is_char_model:
+        return ' '.join([id2word[x] for x in items])
+    else:
+        return ''.join([id2word[x] for x in items]).replace('_', ' ')
+
+
+###############################################################################
+# Input words
+# ---------------------------------------------
```
Comment (srkreddy1238) [inline tutorials/nnvm/from_tensorflow_rnn.py]: under line up to the text.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 7]
```diff
@@ -469,6 +469,224 @@ def _impl(inputs, attr, params):
...
+        i = sigmoid(cs_prev * wci + i)
+        f = sigmoid(cs_prev * wcf + f)
+        ci = tanh(ci)
+
+        cs = ci .* i + cs_prev .* f
+        cs = clip(cs, cell_clip)
+
+        o = sigmoid(cs * wco + o)
+        co = tanh(cs)
+        h = co .* o
```
Comment (Huyuwei) [inline nnvm/python/nnvm/frontend/tensorflow.py]: can remove these math equations and give a link to https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/rnn/python/ops/lstm_ops.py#L41-L114

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 8]
```diff
@@ -505,9 +723,203 @@ def _impl(inputs, attr, params):
...
+_convert_map_rnn = {
+    'LSTMBlockCell'                     : _LSTMBlockCell(),
 }
 
 
+class RecurrentNetworks(object):
+    """Recurrent network layer handlers.
+
+    Unlike normal operators, recurrent network operators have layer concept.
+    Same operators will be called multiple times (based on number of layers)
```
Comment (Huyuwei) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Does layer refer to steps? or multiple cells (layers) in stacked rnn? The doc here is not very clear.

**Comment by Huyuwei** [comment]
@joyalbin Seems current implementation only supports step=1 and users need to do unrolling manually in the case of multiple steps. Will it be good to add a wrapper to support multiple steps?

@merrymercy may have some suggestions.

**Comment by joyalbin** [comment]
@Huyuwei , @srkreddy1238 I have reworked on all the review comments and updated the PR

@Huyuwei  Current implementation support step=1 at a time. This implementation is similar to tensorflow. Here the model input is one 'word', LSTM input is calculated from the 'word's token. so couldnt able to remove the unrolling.
Can we merge this PR with step=1 and multiple step handling in the next PR? Could you please put more light on avoiding this unrolling to handle multiple steps?

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 8 above)
Comment (joyalbin) [reply nnvm/python/nnvm/frontend/tensorflow.py]: @Huyuwei here, layers refers to multiple cells in RNN stack. I have rephrased the comments

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 1 above)
Comment (joyalbin) [reply nnvm/python/nnvm/frontend/tensorflow.py]: @srkreddy1238 The frontend is using NNVM stridedslice only. But tensorflow have additional mask attributes for stridedslice. This is handled here.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 2 above)
Comment (joyalbin) [reply nnvm/python/nnvm/frontend/tensorflow.py]: _infer_out_shapes is used for finding intermediate node shapes

**Comment by joyalbin** [comment]
@Huyuwei , @srkreddy1238 can you please help to review this PR further?

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 9]
```diff
@@ -505,9 +701,198 @@ def _impl(inputs, attr, params):
...
+    'LSTMBlockCell'                     : _LSTMBlockCell(),
 }
 
 
+class RecurrentNetworks(object):
+    """Recurrent network layer handlers.
+
+    Unlike normal operators, stacked rnn have cells and layer concepts.
+    Each Layer represent a cell in RNN stack. Cells in the same RNN stack
+    sequentialy process input data.
```
Comment (Huyuwei) [inline nnvm/python/nnvm/frontend/tensorflow.py]: The description here is still a little confusing. Where is the recurrent part?
Can remove it or give a link to some RNN tutorial.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 10]
```diff
@@ -528,6 +694,120 @@ def test_forward_mobilenet():
...
+    with tf.Graph().as_default():
+        word_to_id, id_to_word, graph_def = nnvm.testing.tf.get_workload_ptb()
+        vocab_size = len(word_to_id)
+        # Call the utility to import the graph definition into default graph.
+        graph_def = nnvm.testing.tf.ProcessGraphDefParam(graph_def)
+        sess = tf.Session()
+    initializer = tf.random_uniform_initializer(config.init_scale,
+                                                config.init_scale)
+    with tf.variable_scope("Model", reuse=None, initializer=initializer):
+        mtest = nnvm.testing.tf.PTBModel(is_training=False, config=config)
```
Comment (Huyuwei) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: mtest is not used? Then testing.tf.PTBModel can be removed

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 11]
```diff
@@ -528,6 +694,120 @@ def test_forward_mobilenet():
...
+
+    #TVM graph module creation
+    params, m = _get_tvm_graph_module(graph_def)
+
+
```
Comment (Huyuwei) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: remove one blank line

**Comment by Huyuwei** [comment]
@joyalbin Sorry for the late response. Have added some comments.

The lstm part looks good to me. fill, gather, stridedslice operators need review from @srkreddy1238

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 12]
```diff
@@ -505,9 +701,198 @@ def _impl(inputs, attr, params):
...
+            Operator name, such as LSTMBlockCell
+
+        inputs : nnvm.Symbol
+            Input data
+
+        attrs : dict
+            Dict of operator attributes
+
+        params : dict
+            List of  pretrained weights and bias
```
Comment (nishi-t) [inline nnvm/python/nnvm/frontend/tensorflow.py]: redundant whitespace after `of`

File: nnvm/python/nnvm/testing/tf.py [hunk 13]
```diff
@@ -257,3 +258,236 @@ def get_workload_mobilenet():
...
+
+    Returns
+    -------
+    graph_def: graphdef
+        graph_def is the tensorflow workload for ptb.
+
+    word_to_id : dict
+        English word to integer id mapping
+    id_to_word : dict
+
```
Comment (nishi-t) [inline nnvm/python/nnvm/testing/tf.py]: Could you remove this blank line?

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 14]
```diff
@@ -505,9 +701,198 @@ def _impl(inputs, attr, params):
...
+            input placeholder.
+
+        inputs : nnvm.Symbol
+            Input data
+
+        attrs : dict
+            Dict of operator attributes
+
+        params : dict
+            List of  pretrained weights and bias
```
Comment (nishi-t) [inline nnvm/python/nnvm/frontend/tensorflow.py]: whitespace after 'of'.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 15]
```diff
@@ -1,4 +1,4 @@
-# pylint: disable=import-self, invalid-name, unused-argument
+# pylint: disable=import-self, invalid-name, unused-argument, too-many-lines
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: too-many-lines ??
Suggest to have logical split instead. If any block can't be reduced use local enable/disable instead.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 16]
```diff
@@ -469,6 +469,211 @@ def _impl(inputs, attr, params):
...
+        begin_input = params.pop(inputs[1].list_output_names()[0]).asnumpy().tolist()
+        end_input = params.pop(inputs[2].list_output_names()[0]).asnumpy().tolist()
+        stride_input = params.pop(inputs[3].list_output_names()[0]).asnumpy().tolist()
+        begin_mask = int(attr.get('begin_mask', 0))
+        end_mask = int(attr.get('end_mask', 0))
+        ellipsis_mask = int(attr.get('ellipsis_mask', 0))
+        new_axis_mask = int(attr.get('new_axis_mask', 0))
+        shrink_axis_mask = int(attr.get('shrink_axis_mask', 0))
+
+        #Constant values used forming output shape.
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Thanks for cleaning the above logic.
Would suggest to clean the below part as well a bit.

My understanding of all these masks as below. Please correct me if I am wrong.

begin, end: Basically to ignore a value from the inputs and use 0 for begin, max for end.
elipsismask: Only one bit will be set which is used to expand begin, end and strides to size of input dimensions by filling max range in between.
newaxis mask: add new axis in the result based on the mask bits.
shringaxis_mask: shrink the axis from result based on the mask bits.

I see this logic can be split into below logically separable blocks.
1: Handle begin and end masks on begin, end
2: Expand begin and end based on elipsis_mask
3: Apply strided_slice operation
4: Apply reshape operation with newaxis and shrink axis 

Also suggest to use python API for converting bitmask into lists of integers.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 17]
```diff
@@ -433,6 +441,155 @@ def test_forward_variable():
...
+        np.testing.assert_allclose(tf_output, tvm_output, atol=1e-5, rtol=1e-5)
+        sess.close()
+
+def test_forward_stridedslice():
+    '''test StridedSlice'''
+    _test_stridedslice((3, 4, 3), [1, 0], [4, 3], [2, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, -1, 0], [4, -5, 3], [2, -1, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, -3, 0], [2, -2, 3], [1, 1, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, 1, 0], [4, 4, 3], [2, 1, 1], 'float32')
+    _test_stridedslice((3, 4, 3), [1, 0], [4, 3], [2, 1],
```
Comment (srkreddy1238) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Please add another testcase with begin, end, ellipsis_mask and new/shrink axis mask with multiple axis.

**Comment by joyalbin** [comment]
@srkreddy1238 yes, your understanding is correct. But handling all the 5 masks based on its priority make things bit lengthy.
I have modified the code based on your comments. 
Mask logic is changed as you suggested and testcases are added.

@Huyuwei I have handled all the review comments. Please help to approve the changes

File: nnvm/python/nnvm/testing/tf.py [hunk 18]
```diff
@@ -134,3 +136,137 @@ def get_workload(model_path):
...
+    samples = []
+    sample = None
+    state_input_name = ['Model/MultiRNNCellZeroState/LSTMBlockCellZeroState/zeros:0',
+                        'Model/MultiRNNCellZeroState/LSTMBlockCellZeroState/zeros_1:0',
+                        'Model/MultiRNNCellZeroState/LSTMBlockCellZeroState_1/zeros:0',
+                        'Model/MultiRNNCellZeroState/LSTMBlockCellZeroState_1/zeros_1:0']
+    state = session.run(state_input_name)
+    fetches = [['Model/RNN/RNN/multi_rnn_cell/cell_0/lstm_cell/LSTMBlockCell:1',
+                'Model/RNN/RNN/multi_rnn_cell/cell_0/lstm_cell/LSTMBlockCell:6',
+                'Model/RNN/RNN/multi_rnn_cell/cell_0/lstm_cell/LSTMBlockCell_1:1',
```
Comment (Huyuwei) [inline nnvm/python/nnvm/testing/tf.py]: Could you add a comment here? What is fetches? What is in the content of LSTMBlockCell:6?

File: nnvm/python/nnvm/testing/tf.py [hunk 19]
```diff
@@ -134,3 +136,137 @@ def get_workload(model_path):
...
+    """Identify token from Softmax output.
+    This token will be mapped to word in the vocabulary.
+    """
+    weight = weight**pows
+    t = np.cumsum(weight)
+    s = np.sum(weight)
+    return int(np.searchsorted(t, 0.5 * s))
+
+def do_tf_sample(session, data, in_states, num_samples):
+    """Sampled from the model"""
```
Comment (Huyuwei) [inline nnvm/python/nnvm/testing/tf.py]: ``data`` is not used in the function.

Comment (joyalbin) [reply nnvm/python/nnvm/testing/tf.py]: @Huyuwei `data `is used at line 192, could you please correct me if I didn't get the real issue here

Comment (Huyuwei) [reply nnvm/python/nnvm/testing/tf.py]: @joyalbin sorry, my wrong.

**Comment by tqchen** [comment]
Thanks @joyalbin @Huyuwei @nishi-t @srkreddy1238 , this is merged

**Pull request merged**